content/post/83-go-ipfs-0.4.20.md

```diff
@@ -0,0 +1,492 @@
+
+All of the resolve commands now:
+
+1. Resolve _recursively_ (up to 32 steps) by default to better match user
+   expectations (these commands used to be non-recursive by default). To turn
+   recursion off, pass `-r false`.
+2. When resolving non-recursively, these commands no longer fail when partially
+   resolving a name. Instead, they simply return the intermediate result.
+
+### Changed: `ipfs files flush`
+
+The `ipfs files flush` command now returns the CID of the flushed file.
+
+## Performance And Reliability
+
+This release has the usual collection of performance and reliability
+improvements.
+
+### 🐺 Badger Memory Usage
+
+Those of you using the badger datastore should notice reduced memory usage in
+this release due to some upstream changes. Badger still uses significantly more
+memory than the default datastore configuration but this will hopefully continue
+to improve.
+
+### 🔁 Bitswap
+
+We fixed some critical CPU utilization regressions in bitswap for this release.
+If you've been noticing CPU _regressions_ in go-ipfs 0.4.19, especially when
+running a public gateway, upgrading to 0.4.20 will likely fix them.
+
+### 🏃‍ Relays
+
+After AutoRelay was introduced in go-ipfs 0.4.19, the number of peers connecting
+through relays skyrocketed to over 120K concurrent peers. This highlighted some
+performance issues that we've now fixed in this release. Specifically:
+
+* We've significantly reduced the amount of memory allocated per-peer.
+* We've fixed a bug where relays might, in rare cases, try to actively dial a
+  peer to relay traffic. By default, relays only forward traffic between peers
+  already connected to the relay.
+* We've fixed quite a number of performance issues that only show up when
+  rapidly forming new connections. This will actually help _all_ nodes but will
+  especially help relays.
+  
+If you've enabled relay _hop_ (`Swarm.EnableRelayHop`) in go-ipfs 0.4.19 and it
+hasn't burned down your machine yet, this release should improve things
+significantly. However, relays are still under heavy load so running an open
+relay will continue to be resource intensive.
+
+We're continuing to investigate this issue and have a few more patches on the
+way that, unfortunately, won't make it into this release.
+
+### 😱 Panics
+
+We've fixed two notable panics in this release:
+
+* We've fixed a frequent panic in the DHT.
+* We've fixed an occasional panic in the experimental QUIC transport.
+
+## 📖 Content Routing
+
+IPFS announces and finds content by sending and retrieving content routing
+("provider") records to and from the DHT. Unfortunately, sending out these
+records can be quite resource intensive.
+
+This release has two changes to alleviate this: a reduced number of initial
+provide workers and a persistent provider queue.
+
+We've reduced the number of parallel initial provide workers (workers that send
+out provider records when content is initially added to go-ipfs) from 512 to 6.
```

Comment:
```suggestion
out provider records when content is initially added to go-ipfs) from 512 to 6, to avoid additional performance overhead from simultaneous provide requests.
```